torch/_library/define.py

```diff
@@ -0,0 +1,334 @@
+            >>>     assert x.dim() == 2
+            >>>     assert weight.dim() == 2
+            >>>     assert bias.dim() == 1
+            >>>     assert x.shape[1] == weight.shape[1]
+            >>>     assert weight.shape[0] == bias.shape[0]
+            >>>     assert x.device == weight.device
+            >>>     return x.new_empty(x.size(0), weight.size(0))
+            >>>
+            >>> x = torch.randn(2, 2)
+            >>> weight = torch.randn(2, 2)
+            >>> bias = torch.randn(2)
+            >>> # xdoctest: +SKIP("Requires Python <= 3.11")
+            >>> out = torch.compile(custom_linear, fullgraph=True)(x, weight, bias)
+            >>> assert torch.allclose(out, torch.nn.functional.linear(x, weight, bias))
+            >>>
+            >>> # Example 2: an operator with data-dependent output shape
+            >>> @torch.library.opaque_op(mutated_args=())
+            >>> def custom_nonzero(x: Tensor) -> Tensor:
+            >>>     x_np = x.cpu().numpy()
+            >>>     res = np.stack(np.nonzero(x_np), axis=1)
+            >>>     return torch.tensor(res, device=x.device)
+            >>>
+            >>> @custom_nonzero.impl_abstract
+            >>> def _(x):
+            >>>     # Number of nonzero-elements is data-dependent.
+            >>>     # Since we cannot peek at the data in an abstract impl,
+            >>>     # we use the ctx object to construct a new symint that
+            >>>     # represents the data-dependent size.
+            >>>     ctx = torch.library.get_ctx()
+            >>>     nnz = ctx.new_dynamic_size()
+            >>>     shape = [nnz, x.dim()]
+            >>>     result = x.new_empty(shape, dtype=torch.int64)
+            >>>     return result
+            >>>
+            >>> x = torch.tensor([0, 1, 2, 0, 0, 1])
+            >>> # xdoctest: +SKIP("Requires Python <= 3.11")
+            >>> out = torch.compile(custom_nonzero)(x)
+            >>> assert torch.allclose(out, x.nonzero())
+
+        """
+        self._abstract_fn = fn
+        return fn
+
+    def _build(self):
+        lib = self._lib
+        lib.define(f"{self._name}{self._schema}")
+        self._opoverload = _library.utils.lookup_op(self._qualname)
+
+        def abstract_impl(*args, **kwargs):
+            if self._abstract_fn is None:
+                import torch
+
+                raise torch._subclasses.fake_tensor.UnsupportedOperatorException(
+                    self._opoverload
+                )
+            return self._abstract_fn(*args, **kwargs)
+
+        library.impl_abstract(self._qualname, lib=lib)(abstract_impl)
+
+    def __call__(self, *args, **kwargs):
+        return self._opoverload(*args, **kwargs)
+
+
+OPDEF_TO_LIB: Dict[str, "library.Library"] = {}
+
+
+def get_library_allowing_overwrite(namespace: str, name: str) -> "library.Library":
+    qualname = f"{namespace}::{name}"
+
+    if qualname in OPDEF_TO_LIB:
+        OPDEF_TO_LIB[qualname]._destroy()
```

Comment:
🤔

Reply:
Is the reason you do this because, even though you mangle modules, you absolutely do not want to mangle names? But if the modules are mangled, there isn't really any expectation of name stability, and I feel you should be willing to mangle names.

If someone wants to rely on the exact naming of the custom operator, then we should force the user to explicitly say what they want the name to be. In fact, it may be worthwhile to make it a lot *harder* for people to try to manually reverse engineer the mangling scheme, because the current module mangling means that transforms like moving a function from one file to another will change its name and would then be BC-breaking. One way to make it hard is to put a big ass string in the module name being like "DONT_USE_THIS_GIVE_EXPLICIT_NAME_IF_NEEDED"

Reply:
Updated the mangling scheme as per your comments